sans/direct-beam.ipynb

```diff
@@ -0,0 +1,529 @@
+    "    # Use a mask, so we can just use `sc.mean`.\n",
+    "    # Note that in the original script Integration(...)/delta_Q is used, but this\n",
+    "    # is biased due to log binning, maybe on purpose...?\n",
+    "    inv_w = sc.reciprocal(wavelength)\n",
+    "    dim = inv_w.dims[0]\n",
+    "    d_inv_w = sc.concatenate(inv_w[dim,1:], inv_w[dim,:-1], 'bound') - inv_w[dim,-1]\n",
+    "    q_range = qlongW + d_inv_w * (qshortW - qlongW) / (inv_w[dim,0] - inv_w[dim,-1])\n",
+    "    qmin = q_range['bound',0]\n",
+    "    qmax = q_range['bound',1]\n",
+    "    return sc.greater(qmin, q_bins['Q',:-1]) | sc.less(qmax, q_bins['Q',1:])\n",
+    "\n",
+    "def interpolate_cubic_spline(data, x, dim):\n",
+    "    import functools\n",
+    "    from scipy import interpolate\n",
+    "    def interpolate_layer(i):\n",
+    "        tck = interpolate.splrep(\n",
+    "            x=contrib.midpoints(data.coords[dim], dim).values,\n",
+    "            y=data['layer',i].values,\n",
+    "            w=1.0/np.sqrt(data['layer',i].variances))\n",
+    "        # TODO uncertainties\n",
+    "        return sc.Variable(dims=[dim], values=interpolate.splev(x.values, tck))\n",
+    "    y = map(interpolate_layer, range(data.coords['layer'].shape[0]))\n",
+    "    concat = functools.partial(sc.concatenate, dim='layer')\n",
+    "    return sc.DataArray(data=functools.reduce(concat, y), coords={dim:x})\n",
+    "\n",
+    "def to_q_by_wavelength(data, transmission, direct_beam, direct_beam_transmission, masks, wavelength_bands):\n",
+    "    wav = sans.to_wavelength(data=data,\n",
+    "                             transmission=transmission,\n",
+    "                             direct_beam=direct_beam,\n",
+    "                             direct_beam_transmission=direct_beam_transmission,\n",
+    "                             masks=masks,\n",
+    "                             wavelength_bins=wavelength_bins)\n",
+    "    return sans.reduce_by_wavelength(wav, q_bins, groupby='layer', wavelength_bands=wavelength_bands)\n",
+    "\n",
+    "def direct_beam_iteration(direct_beam, layers, masks, wavelength_bands, flat_background=0.25*sc.units.one):\n",
+    "    print('start iteration')\n",
+    "    direct_beam_by_pixel = sc.choose(layers, choices=direct_beam, dim='layer')\n",
+    "    sample_q_lambda = to_q_by_wavelength(data=sample,\n",
+    "                                         transmission=sample_trans,\n",
+    "                                         direct_beam=direct_beam_by_pixel,\n",
+    "                                         direct_beam_transmission=background_trans, # note: background_trans\n",
+    "                                         masks=masks,\n",
+    "                                         wavelength_bands=wavelength_bands)\n",
+    "    background_q_lambda = to_q_by_wavelength(data=background,\n",
+    "                                             transmission=background_trans,\n",
```

Comment:
variables such as `background_trans` from global scope, should be passed as parameters.